### PR TITLE
Native TPC-H Database Generator

### PR DESCRIPTION
**Pull request by bluecat1024**
native generation for TPC-H workload

### CODE REVIEW

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 1]
```diff
@@ -213,21 +168,11 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of Part because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO part " + "(p_partkey, p_name, p_mfgr, p_brand, p_type," + " p_size, p_container, p_retailprice, p_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:171]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 2]
```diff
@@ -169,38 +146,16 @@ private boolean loadCopy(Connection conn) {
...
+                    List<Iterable<List<Object>>> regionGenerators = new ArrayList<>();
+                    regionGenerators.add(new RegionGenerator());
 
-            @Override
-            public void beforeLoad() {
-                try {
-                    copyLatch.await();
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
+                    genTable(conn, statement, regionGenerators, regionTypes, "Region");
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:158]: Don't use hardcoded strings for table names. Define them in `TPCHConstants`.

https://github.com/cmu-db/benchbase/blob/main/src/main/java/com/oltpbenchmark/benchmarks/tatp/TATPConstants.java#L39-L52

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 3]
```diff
@@ -240,19 +185,17 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of Nation because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO nation " + "(n_nationkey, n_name, n_regionkey, n_comment) " + "VALUES (?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:188]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 4]
```diff
@@ -268,19 +211,17 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of Supplier because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO supplier " + "(s_suppkey, s_name, s_address, s_nationkey, s_phone," + " s_acctbal, s_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:214]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 5]
```diff
@@ -296,19 +237,17 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of Customer because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO customer " + "(c_custkey, c_name, c_address, c_nationkey," + " c_phone, c_acctbal, c_mktsegment, c_comment ) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:240]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 6]
```diff
@@ -324,19 +263,17 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of Orders because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO orders " + "(o_orderkey, o_custkey, o_orderstatus, o_totalprice," + " o_orderdate, o_orderpriority, o_clerk, o_shippriority," + " o_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:266]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 7]
```diff
@@ -352,19 +289,17 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of PartSupp because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO partsupp " + "(ps_partkey, ps_suppkey, ps_availqty, ps_supplycost," + " ps_comment) " + "VALUES (?, ?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:292]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java [hunk 8]
```diff
@@ -381,19 +316,17 @@ public void afterLoad() {
         threads.add(new LoaderThread(this.benchmark) {
             @Override
             public void load(Connection conn) throws SQLException {
-                if (copySuccess[0]) {
-                    LOG.info("Skipping LOAD of LineItem because COPY has been done.");
-                    return;
-                }
                 try (PreparedStatement statement = conn.prepareStatement("INSERT INTO lineitem " + "(l_orderkey, l_partkey, l_suppkey, l_linenumber," + " l_quantity, l_extendedprice, l_discount, l_tax," + " l_returnflag, l_linestatus, l_shipdate, l_commitdate," + " l_receiptdate, l_shipinstruct, l_shipmode, l_comment) " + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/TPCHLoader.java:319]: Use `SQLUtil.getInsertSQL`

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.oltpbenchmark.benchmarks.tpch.generation;
+
+import static com.google.common.base.Preconditions.checkState;
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java:16]: Does any other part of BenchBase use this? If no, we should remove it.

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
+/*
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java:1]: If this code came from the Trino, add their license info to make it clear we got it from them. Do this for all the files that we copy.

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.oltpbenchmark.benchmarks.tpch.generation;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public abstract class AbstractRandomInt
+{
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/AbstractRandomInt.java:19]: Make sure the new code matches the same code style as the rest of the repository.

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+package com.oltpbenchmark.benchmarks.tpch.generation;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
```
Comment (apavlo) [inline src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java:23]: We should probably remove this.

File: src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss [hunk 13]
```diff
@@ -0,0 +1,836 @@
+#
```
Comment (apavlo) [inline src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss:1]: @timveil Should this file be in `src/main/resources/benchmarks/tpch`?

File: src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
+import java.util.List;
+
+import junit.framework.TestCase;
+
+import com.oltpbenchmark.benchmarks.tpch.generation.*;
+
+public class TestGenerators extends TestCase {
```
Comment (apavlo) [inline src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java:7]: Rename to `TestTPCHGenerators`

File: src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+    public TestGenerators() {
+        super();
+    }
+
+    public void testGenerators() {
```
Comment (apavlo) [inline src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java:13]: Are you just printing the output of the generators? How are we going to check whether the generators contain correct data when the test runs automatically in Github?

Also if the default scalefactor is one, does this mean that we're going to generate 1GB of debug output?

**Review by apavlo — CHANGES_REQUESTED** [review]
Fix the addressed comments.
Rename `tpch/generation` -> `tpch/util` to match other benchmarks.

Instead of a custom random data generators, I would like to maybe try to combine what we already have with `benchbase/src/main/java/com/oltpbenchmark/util/RandomGenerator.java`. The TPC-H random generator is more advanced, so we should merge their code into our system. But we can will want to get rid of the `expectedUsagePerRow` hint. We use `RandomGenerator` to create input parameters for txns so there isn't a notion of a row in that scenario.

File: src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java (hunk 15 above)
Comment (bluecat1024) [reply src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java:13]: For automatic check I should use assertEqual with some ground truth data.

In those output I hardcoded those SFs to be very small. SFs are parameters for generators.

Comment (apavlo) [reply src/test/java/com/oltpbenchmark/benchmarks/tpch/generation/TestGenerators.java:13]: There is already testcase infrastructure for automatically testing benchmark data loaders. Make a new class `TestTPCHLoader` that extends `AbstractTestLoader`:

Examples:
* https://github.com/cmu-db/benchbase/blob/main/src/test/java/com/oltpbenchmark/benchmarks/tpcc/TestTPCCLoader.java
* https://github.com/cmu-db/benchbase/blob/main/src/test/java/com/oltpbenchmark/benchmarks/tatp/TestTATPLoader.java
* thttps://github.com/cmu-db/benchbase/blob/main/src/test/java/com/oltpbenchmark/benchmarks/tpcc/TestTPCCLoader.java

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java (hunk 12 above)
Comment (bluecat1024) [reply src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java]: I am not sure I have resolved this. Do you mean only the checkState and toStringHelper should be removed, or all com.google.common should be avoided?

File: src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss (hunk 13 above)
Comment (bluecat1024) [reply src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss]: Shall this be re-located?

**Comment by bluecat1024** [comment]
> Fix the addressed comments. Rename `tpch/generation` -> `tpch/util` to match other benchmarks.
> 
> Instead of a custom random data generators, I would like to maybe try to combine what we already have with `benchbase/src/main/java/com/oltpbenchmark/util/RandomGenerator.java`. The TPC-H random generator is more advanced, so we should merge their code into our system. But we can will want to get rid of the `expectedUsagePerRow` hint. We use `RandomGenerator` to create input parameters for txns so there isn't a notion of a row in that scenario.

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java (hunk 12 above)
Comment (apavlo) [reply src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java]: Do we already include `com.google.common` packages in maven? I don't want additional dependencies just for these checks.

File: src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss (hunk 13 above)
Comment (apavlo) [reply src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss]: Yes.

File: src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java (hunk 12 above)
Comment (bluecat1024) [reply src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java]: Not changing the pom.xml. The repo compiles.

The pom has a com.google.cloud pakage which fetches all google.common so no additional dependency needed.
Yet strangely, google packages are used nowhere except for the TPCH code I have introduced.

Comment (bluecat1024) [reply src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java]: So the google package functionalities may be replaced by self-defined ones, and com.google.cloud can be instead deleted (google dependency tree is big).

Comment (bluecat1024) [reply src/main/java/com/oltpbenchmark/benchmarks/tpch/generation/Distribution.java]: resolved this. no google pkg needed.

File: src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss (hunk 13 above)
Comment (bluecat1024) [reply src/main/resources/com/oltpbenchmark/benchmarks/tpch/generation/dists.dss]: resolved

**Comment by bluecat1024** [comment]
> Fix the addressed comments. Rename `tpch/generation` -> `tpch/util` to match other benchmarks.
> 
> Instead of a custom random data generators, I would like to maybe try to combine what we already have with `benchbase/src/main/java/com/oltpbenchmark/util/RandomGenerator.java`. The TPC-H random generator is more advanced, so we should merge their code into our system. But we can will want to get rid of the `expectedUsagePerRow` hint. We use `RandomGenerator` to create input parameters for txns so there isn't a notion of a row in that scenario.

The `expectedUsagePerRow` is a concept for RandomGenerators to skip a certain number of rows and columns, which is often used the native generation of TPCH and TPCDS. So I suppose this part of logic can be refactored as another "RandomInt" in the util package, so both loaders share the logic of it.

**Pull request merged**